---
title: "R Notebook"
output: html_notebook
---
Perform custom contrasts to explore levels not compared in main models.
* this increases power i think, relative to straight-forward pairwise comparison

Create combined age-habitat factor and sex-habitat
* simple effect [see: http://glimo.vub.ac.be/downloads/simpleeffect.htm]
```{r Housekeeping, echo=FALSE}
rm(list=ls())
#set.seed(10)
```

```{r, eval=FALSE, results='hide', warning=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")
BiocManager::install("metagenomeSeq")
BiocManager::install("microbiome")

```

```{r Set-up, message=FALSE, echo=FALSE}
# libraries
#install.packages("tidyverse", dependencies = T)
library(vegan)  ##note had trouble installing, needed to install gfortran first
library(tidyverse) #
library(phyloseq) 
library(lme4)
library(microbiome)
library(gridExtra)
library(ggpubr)
library(lmerTest) # NOTE: overloads lme4 and extends test to provide p-values AND therefore causes DHARMa to give a warning but doesnt appear to effect it
library(DHARMa)
library(kableExtra)
#library(sjPlot) # do i trust this package? Use effects instead.
library(effects)
library(emmeans)
```

```{r Read in data, warning=FALSE, echo=FALSE}
phylo.spring <- readRDS(file = "../Data/phylo-spring.rds")
phylo.spring
```

```{r}
sample_data(phylo.spring)$age.habitat <- as.factor(paste(sample_data(phylo.spring)$ageBinned, sample_data(phylo.spring)$habitat, sep = "."))
#metadata <- cbind(metadata, age.habitat)

# sex.habitat <- paste(metadata$Sex, metadata$habitat, sep = ".")
# metadata <- cbind(metadata, sex.habitat)
```

```{r}
metadata <- meta(phylo.spring) # doesnt maintain Date data-type

numeric.predictors <- c("QubitDNA","Tarsus","Weight", "wing", "broodSizeWhenSampled", "broodSizeMax", "totalFledge", "clutchSize", "layDateFirst", "numberDeadPreRinged", "numberDeadPostRinged","scaled.mass","scaled.mass.wing.adult", "scaled.mass.tarsus.adult","scaled.mass.chick", "scaled.mass.tarsus", "scaled.mass.wing", "DistanceToEdge")

# centre and scale numeric variables ie. subtract mean and divide by st. deviation
metadata.scaled <- metadata
metadata.scaled[,numeric.predictors] <- scale(metadata.scaled[,numeric.predictors])
```

```{r}
# Split data into adult and nestlings. Use ageBinned("1week","2week","adult").
# subset to nestlings
## alt use meta() to extract metadata as df
phylo.nestlings <- subset_samples(phylo.spring, ageBinned=="1week"|ageBinned=="2week")
nestlings.meta <- meta(phylo.nestlings@sam_data)

# subset to adults
phylo.adults <- subset_samples(phylo.spring, ageBinned=="adult")
adults.meta <- meta(phylo.adults@sam_data)

adults.meta.scaled <- adults.meta
adults.meta.scaled[,numeric.predictors] <- scale(adults.meta.scaled[,numeric.predictors])

nestlings.meta.scaled <- nestlings.meta
nestlings.meta.scaled[,numeric.predictors] <- scale(nestlings.meta.scaled[,numeric.predictors])
```


# All birds

## Shannon


```{r}
ggplot(metadata, aes(age.habitat, shannon, color=age.habitat)) +
  geom_boxplot() +
 # geom_jitter(width = 0.2) +
  ggtitle("Shannon diversity")
```

Run model w/ new combined factor
```{r}
metadata$age.habitat <- relevel(metadata$age.habitat, "1week.deciduous")
all.sh.lmm <- lmer(log(shannon) ~ age.habitat + (1|site/nest/bird.ID) + (1|SequencePlate), data = metadata)
summary(all.sh.lmm)
```

Emmeans
```{r}
emm1 = emmeans(all.sh.lmm, specs = ~ age.habitat, adjust="tukey")
emm1
```

Emmeans custom contrasts
* only making contrast that arent made in the main model in order to increase power
```{r}
week1.decid = c(1, 0, 0, 0, 0, 0)
week1.conif = c(0, 1, 0, 0, 0, 0)
week2.conif = c(0, 0, 1, 0, 0, 0)
week2.decid = c(0, 0, 0, 1, 0, 0)
adult.conif = c(0, 0, 0, 0, 1, 0)
adult.decid = c(0, 0, 0, 0, 0, 1)


contrast(emm1, method = list("week2.decid - adult.decid" = week2.decid - adult.decid,
                             "week2.decid - week1.conif" = week2.decid - week1.conif,
                             "week2.decid - week2.conif" = week2.decid - week2.conif,
                             "week2.decid - adult.conif" = week2.decid - adult.conif,
                             "adult.decid - week1.conif" = adult.decid - week1.conif,
                             "adult.decid - week2.conif" = adult.decid - week2.conif,
                             "adult.decid - adult.conif" = adult.decid - adult.conif,
                             "week1.conif - week2.conif" = week1.conif - week2.conif,
                             "week1.conif - adult.conif" = week1.conif - adult.conif,
                             "week2.conif - adult.conif" = week2.conif - adult.conif) )

```

## Chao1

```{r}
#all.chao.lmm <- lmer(log(chao1) ~ age.habitat +(1|SequencePlate) +(1|site/nest/bird.ID) , data = metadata)
summary(all.chao.lmm)
```

```{r}
emm2 = emmeans(all.chao.lmm, specs = ~ age.habitat, adjust="tukey")
emm2
```

```{r}
contrast(emm2, method = list("week2.decid - adult.decid" = week2.decid - adult.decid,
                             "week2.decid - week1.conif" = week2.decid - week1.conif,
                             "week2.decid - week2.conif" = week2.decid - week2.conif,
                             "week2.decid - adult.conif" = week2.decid - adult.conif,
                             "adult.decid - week1.conif" = adult.decid - week1.conif,
                             "adult.decid - week2.conif" = adult.decid - week2.conif,
                             "adult.decid - adult.conif" = adult.decid - adult.conif,
                             "week1.conif - week2.conif" = week1.conif - week2.conif,
                             "week1.conif - adult.conif" = week1.conif - adult.conif,
                             "week2.conif - adult.conif" = week2.conif - adult.conif) )

```

# Adults
Should i perform this simple effect comparison when there is no evidence of an interaction?
## Shannon
```{r}
# DistanceToEdge + broodSizeMax + layDateFirst +
ad.sh.lmm <- lmer((shannon)^(1/3) ~ Sex * habitat +  (1|SequencePlate) + (1|site/nest), data = adults.meta.scaled)
summary(ad.sh.lmm)
```

## Chao1
```{r}
# no adult model using emmeans
```

# Nestling
```{r}
table(nestlings.meta.scaled$age.habitat)
```

## Shannon
 
```{r}
nestlings.meta.scaled$age.habitat <- relevel(nestlings.meta.scaled$age.habitat, "1week.deciduous")
nstl.sh.lmm <- lmer(shannon^(1/3)~ age.habitat + DistanceToEdge + layDateFirst +(1|SequencePlate) + (1|site/nest/bird.ID), data = nestlings.meta.scaled)

summary(nstl.sh.lmm)
```
Note: didnt run when shannon in brackets
```{r}
emm5 = emmeans(nstl.sh.lmm, specs = ~ age.habitat, adjust="tukey")
emm5
```

```{r}
week1.decid = c(1, 0, 0, 0)
week1.conif = c(0, 1, 0, 0)
week2.conif = c(0, 0, 1, 0)
week2.decid = c(0, 0, 0, 1)

contrast(emm5, method = list("week1.conif - week2.conif" = week1.conif - week2.conif,
                             "week1.conif - week2.decid" = week1.conif - week2.decid,
                             "week2.conif - week2.decid" = week2.conif - week2.decid), by = NULL )
```

## Chao1

```{r}
nstl.chao.lmm <- lmer(log(chao1)~ age.habitat + broodSizeWhenSampled + DistanceToEdge + layDateFirst + (1|SequencePlate) + (1|site/nest/bird.ID), data = nestlings.meta.scaled)
summary(nstl.chao.lmm)
```


```{r}
emm6 = emmeans(nstl.chao.lmm, specs = ~ ageBinned:habitat, adjust="tukey")
emm6
```

```{r}
contrast(emm6, method = list("week2.decid - week1.conif" = week2.decid - week1.conif,
                             "week2.decid - week2.conif" = week2.decid - week2.conif,
                             "week1.conif - week2.conif" = week1.conif - week2.conif) )

```
